### PR TITLE
overlord/configstate: support nested configuration

### PR DESCRIPTION
**Pull request by niemeyer**

### CODE REVIEW

File: overlord/configstate/transaction.go [hunk 1]
```diff
@@ -114,7 +152,106 @@ func (t *Transaction) GetMaybe(snapName, key string, result interface{}) error {
...
+	case nil:
+		// Missing update map. Create and nest final value under it.
+		configm := make(map[string]interface{})
+		_, err := patchConfig(snapName, subkeys, pos, configm, value)
+		if err != nil {
+			return nil, err
+		}
+		return configm, nil
+
+	case *json.RawMessage:
```
Comment (pedronis) [inline overlord/configstate/transaction.go]: this happens when one of the value set was a map itself?

File: overlord/configstate/transaction.go [hunk 2]
```diff
@@ -168,16 +333,3 @@ type NoOptionError struct {
 func (e *NoOptionError) Error() string {
 	return fmt.Sprintf("snap %q has no %q configuration option", e.SnapName, e.Key)
```
Comment (niemeyer) [inline overlord/configstate/transaction.go]: We test NoOptionError, but not the message. Will add the test.

File: overlord/configstate/transaction.go [hunk 3]
```diff
@@ -114,7 +152,106 @@ func (t *Transaction) GetMaybe(snapName, key string, result interface{}) error {
...
+			return fmt.Errorf("internal error: cannot unmarshal snap %q option %q into %T: %s, json: %s", snapName, key, result, err, *raw)
+		}
+		return nil
+	}
+
+	configm, ok := value.(map[string]interface{})
+	if !ok {
+		raw, ok := value.(*json.RawMessage)
+		if !ok {
+			raw = jsonRaw(value)
```
Comment (niemeyer) [inline overlord/configstate/transaction.go]: I don't think we can reach it with the current logic.

File: overlord/configstate/transaction.go (hunk 1 above)
Comment (niemeyer) [inline overlord/configstate/transaction.go]: It should happen on any final value. Anything set=<value>, value is marshaled before going into the tree.

Per the logic below (and tests), it breaks politely if the value is not a map though.

**Pull request merged**